### PR TITLE
[ENG-1553] Institutions Summary Endpoint

### PR DESCRIPTION
**Pull request by UdayVarkhedkar**
## Purpose
To create an API endpoint which may be used to retrieve a summary of Institutional metrics including the total number of public projects, private projects, and users affiliated with a given institution.

## Changes
* Creation of InstitutionMetricsSummary endpoint in `api/institutions/metrics/`
* Addition of user_count to InstitutionProjectCounts metric

### Sample API Output
```
{
    "data": {
        "id": "cos",
        "type": "institution-summary-metrics",
        "attributes": {
            "public_project_count": 24,
            "private_project_count": 27,
            "user_count": 4
        },
        "links": {
            "self": "http://localhost:8000/v2/institutions/cos/metrics/"
        }
    },
    "meta": {
        "version": "2.0"
    }
}
```

## QA Notes
This may require dev testing as the best way to test it is to populate ES6 (Impact) with fake data, make a query to the API endpoint and validate that the data lines up.

## Documentation
If we have internal documentation for the API, we should update that with details on this endpoint.

## Side Effects
N/A

## Ticket
[JIRA Ticket](https://openscience.atlassian.net/browse/ENG-1553)

### CODE REVIEW

File: api/metrics/views.py [hunk 1]
```diff
@@ -151,3 +152,33 @@ def metric_type(self):
     @property
     def metric(self):
         return PreprintDownload
+
+class InstitutionSummaryMetrics(JSONAPIBaseView):
+    # TODO: Identify necessary permissions level for InstitutionSummaryMetrics
```
Comment (Johnetordoff) [inline api/metrics/views.py]: I think we'd be safe with guardian permissions [like so](https://github.com/CenterForOpenScience/osf.io/pull/9318/files#diff-3b3e64edb1e8e2455d8c6c9ba66efe21R153-R164)

File: api/metrics/urls.py [hunk 2]
```diff
@@ -7,4 +7,5 @@
 urlpatterns = [
     url(r'^preprints/views/$', views.PreprintViewMetrics.as_view(), name=views.PreprintViewMetrics.view_name),
     url(r'^preprints/downloads/$', views.PreprintDownloadMetrics.as_view(), name=views.PreprintDownloadMetrics.view_name),
+    url(r'^institutions/$', views.InstitutionSummaryMetrics.as_view(), name=views.InstitutionSummaryMetrics.view_name),
```
Comment (Johnetordoff) [inline api/metrics/urls.py]: @brianjgeiger Is this metrics or institutions endpoint? These are metrics views, but are I made the department endpoint [off institutions](https://github.com/CenterForOpenScience/osf.io/pull/9318/files#diff-e8a52b87015f3732c0abf6916bc414d4R16)

Comment (brianjgeiger) [reply api/metrics/urls.py]: I think they should be institutions endpoints. They'll need to be fully JSON-API compliant and used by our front-end, so no need to put them in the `/_/metrics` namespace.

File: api_tests/institutions/views/test_institution_summary_metrics.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        # Tests unauthorized user
+        resp = app.get(url, auth=user.auth, expect_errors=True)
+        assert resp.status_code == 403
+
+        # Tests authorized user with institution without any metrics
+        resp = app.get(url, auth=admin.auth, expect_errors=True)
+        assert resp.status_code == 404
+
+        # Record latest institutional metrics to ES
+        public_project_count_latest = int(50 * random.random())
```
Comment (Johnetordoff) [inline api_tests/institutions/views/test_institution_summary_metrics.py]: Because if testmon caching there's not really much of a point to randomly generating these, it will just be cached as the same result for subsequent runs. It's also just not really conventionally what we do for tests, we try to use fixed values unless it is data like usernames that have to be unique. 

Also you'd probably want to use `random.randint` instead of this.

File: api/institutions/urls.py [hunk 4]
```diff
@@ -13,5 +13,6 @@
     url(r'^(?P<institution_id>\w+)/relationships/registrations/$', views.InstitutionRegistrationsRelationship.as_view(), name=views.InstitutionRegistrationsRelationship.view_name),
     url(r'^(?P<institution_id>\w+)/relationships/nodes/$', views.InstitutionNodesRelationship.as_view(), name=views.InstitutionNodesRelationship.view_name),
     url(r'^(?P<institution_id>\w+)/users/$', views.InstitutionUserList.as_view(), name=views.InstitutionUserList.view_name),
+    url(r'^(?P<institution_id>\w+)/metrics/$', views.InstitutionSummaryMetrics.as_view(), name=views.InstitutionSummaryMetrics.view_name),
```
Comment (Johnetordoff) [inline api/institutions/urls.py]: This is going to sound stupid, but would you mind changing this to `summary_metrics`. It's just that Corbin's endpoint has to be named `user_metrics` because the `users` endpoint is already taken. This means that we'd have three new enpoints named `user_metrics`, `metrics` and `departments` to avoid confusion, I think it would be best to go with `user_metrics`, `summary_metrics` and `department_metrics`. Which I think is better, but if name any other naming convention ideas, shout out because I'm working on the final relationship ticket now.

Comment (UdayVarkhedkar) [reply api/institutions/urls.py]: Thoughts on `/metrics/` for summary, `/metrics/users` for user metrics, and `/metrics/departments/` for department metrics?

Comment (Johnetordoff) [reply api/institutions/urls.py]: I don't want to make a bare `/metrics/` endpoint in case people think it's a list view of all child metrics endpoints. I wonder if there's some JSON-API convention for this @brianjgeiger , an any ideas?

Comment (brianjgeiger) [reply api/institutions/urls.py]: That seems fine to me.

Comment (Johnetordoff) [reply api/institutions/urls.py]: Ok, the relationships PR will use those endpoints.

Comment (brianjgeiger) [reply api/institutions/urls.py]: I'm going to verify with Front End just to be sure. It may be that ember will want it to be `/metrics/summary` in addition to the other two.

Comment (brianjgeiger) [reply api/institutions/urls.py:16]: @Johnetordoff @UdayVarkhedkar We had a long discussion on the front-end flow, and I think we should go with `/metrics/summary` because the endpoint returns a dict rather than a list, so should be singular.

File: api_tests/institutions/views/test_institution_summary_metrics.py (hunk 3 above)
Comment (UdayVarkhedkar) [reply api_tests/institutions/views/test_institution_summary_metrics.py]: Changed to use specified values

File: api_tests/institutions/views/test_institution_summary_metrics.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        return InstitutionFactory()
+
+    @pytest.fixture()
+    def user(self):
+        return AuthUserFactory()
+
+    @pytest.fixture()
+    def admin(self, institution):
+        user = AuthUserFactory()
+        group = institution.get_group('institutional_admins')
```
Comment (brianjgeiger) [inline api_tests/institutions/views/test_institution_summary_metrics.py:27]: Not part of this PR, but I think there's supposed to be an institutional admin group for each institution. Otherwise anyone who is an admin could view any institution's summary information, which is not appropriate.

Comment (UdayVarkhedkar) [reply api_tests/institutions/views/test_institution_summary_metrics.py:27]: `institution.get_group('institutional_admins')` returns a group in the format `institution_<institution._id>_institutional_admins` in my local shell so the group is specific to the institution which indicates to me that it should be fine - does that sound correct to you?

Comment (brianjgeiger) [reply api_tests/institutions/views/test_institution_summary_metrics.py:27]: Yes, very good. That makes much more sense than what I was thinking first thing this morning. :)
